A50-xds-outlier-detection.md

```diff
@@ -0,0 +1,262 @@
+    // this host. Defaults to 50.
+    google.protobuf.UInt32Value request_volume = 4;
+  }
+
+  // If set, success rate ejections will be performed
+  SuccessRateEjection success_rate_ejection = 5;
+  
+  // If set, failure rate ejections will be performed
+  FailurePercentageEjection failure_percentage_ejection = 6;
+
+  // The config for the child policy
+  repeated LoadBalancingConfig child_policy = 13;
+}
+```
+
+### Outlier Detection Behavior
+
+Internally, the `outlier_detection` LB policy will store a map with keys that are subchannel addresses, and values that are objects that contain the following:
+
+ - The call result counter object (see [Call Counter section](#call-counter))
+ - The latest ejection timestamp, or `null` if the address is currently not ejected
+ - The current ejection time multiplier, starting at 0
+ - A list of subchannel wrapper objects that correspond to this address
+
+When the `outlier_detection` LB policy receives an address update, it will create a map entry for each subchannel address in the list, and remove each map entry for a subchannel address not in the list, then pass the address list along to the subchannel.
+
+When the child policy asks for a subchannel, the `outlier_detection` will wrap the subchannel with a wrapper (see [Subchannel Wrapper section](#subchannel-wrapper)). Then, the subchannel wrapper will be added to the list in the map entry for its address, if that map entry exists. If there is no map entry, or if the subchannel is created with multiple addresses, the subchannel will be ignored for outlier detection. If that address is currently ejected, that subchannel wrapper's `eject` method will be called.
+
+The `outlier_detection` LB policy will provide a picker that delegates to the child policy's picker, and when the request finishes, increment the corresponding counter in the map entry referenced by the subchannel wrapper that was picked.
+
+The `outlier_detection` LB policy will have a timer that triggers on a period determined by the `interval` config option, and does the following:
+
+ 1. For each address, swap the call counter's buckets in that address's map entry.
+ 2. If the `success_rate_ejection` configuration field is set, run the [success rate](#success-rate-algorithm) algorithm.
+ 3. If the `failure_percentage_ejection` configuration field is set, run the [falure percentage](#failure-percentage-algorithm) algorithm.
+ 4. For each address in the map:
+    - If the address is not ejected and the multiplier is greater than 0, decrease the multiplier by 1.
+    - If the address is ejected, and the current time is after `ejection_timestamp + min(base_ejection_time * multiplier, max(base_ejection_time, max_ejection_time))`, un-eject the address.
+
+### Ejection and Un-ejection
+
+To eject an address, set the current ejection timestamp to the current time, increase the ejection time multiplier by 1, and call `eject()` on each subchannel wrapper in that address's subchannel wrapper list.
```

Comment:
Implementation note: You probably don't want to use the current time _now_ when doing this. You probably want to grab the current time when you start the "loop over each address" process. Otherwise you can have aliasing issues where you are 1ms before the ejection window would expire. Grabbing the time at the start of the loop means the time is always _a bit older_ and the comparisons shouldn't offer any surprise.

Alternatively, you could subtract a fixed amount of time (e.g., 1% or 50% of `OutlierDetectionLoadBalancingConfig.interval`) when setting the value.

Reply:
I changed it to use the timestamp of when the timer fires.